group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformAssignor.java

```diff
@@ -0,0 +1,288 @@
+ * @see GeneralUniformAssignmentBuilder
+ */
+public class UniformAssignor implements PartitionAssignor {
+    private static final Logger log = LoggerFactory.getLogger(UniformAssignor.class);
+    public static final String UNIFORM_ASSIGNOR_NAME = "uniform";
+
+    @Override
+    public String name() {
+        return UNIFORM_ASSIGNOR_NAME;
+    }
+
+    /**
+     * Perform the group assignment given the current members and
+     * topic metadata.
+     *
+     * @param assignmentSpec                The member assignment spec.
+     * @param subscribedTopicDescriber      The topic and cluster metadata describer {@link SubscribedTopicDescriber}.
+     * @return The new assignment for the group.
+     */
+    @Override
+    public GroupAssignment assign(
+        AssignmentSpec assignmentSpec,
+        SubscribedTopicDescriber subscribedTopicDescriber
+    ) throws PartitionAssignorException {
+        AbstractAssignmentBuilder assignmentBuilder;
+        if (assignmentSpec.members().isEmpty())
+            return new GroupAssignment(Collections.emptyMap());
+
+        if (allSubscriptionsEqual(assignmentSpec.members())) {
+            log.debug("Detected that all members are subscribed to the same set of topics, invoking the "
+                + "optimized assignment algorithm");
+            assignmentBuilder = new OptimizedUniformAssignmentBuilder(assignmentSpec, subscribedTopicDescriber);
+        } else {
+            assignmentBuilder = new GeneralUniformAssignmentBuilder();
+            log.debug("Detected that all members are subscribed to a different set of topics, invoking the "
+                + "general assignment algorithm");
+        }
+
+        return assignmentBuilder.buildAssignment();
+    }
+
+    /**
+     * Determines if all members are subscribed to the same list of topic IDs.
+     *
+     * @param members A map of member identifiers to their respective {@code AssignmentMemberSpec}.
+     *                Assumes the map is non-empty.
+     * @return true if all members have the same subscription list of topic IDs,
+     *         false otherwise.
+     */
+    private boolean allSubscriptionsEqual(Map<String, AssignmentMemberSpec> members) {
+        Set<Uuid> firstSubscriptionSet = new HashSet<>(members.values().iterator().next().subscribedTopicIds());
+        for (AssignmentMemberSpec memberSpec : members.values()) {
+            Set<Uuid> currentSubscriptionSet = new HashSet<>(memberSpec.subscribedTopicIds());
+            if (!firstSubscriptionSet.equals(currentSubscriptionSet)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * The assignment builder is used to construct the final assignment in a series of steps that
+     * are determined by the type of subscriptions.
+     *
+     * There are common methods present that are used by any type of assignment strategy.
+     */
+     protected static abstract class AbstractAssignmentBuilder {
```

Comment:
Which concrete classes? We made it abstract so that any builder that implements this class would need to implement buildAssignment which is the main function called within assign

Reply:
GeneralUniformAssignmentBuilder and OptimizedUniformAssignmentBuilder. How about extracting AbstractAssignmentBuilder from UniformAssignor and calling it AbstractUniformAssignmentBuilder?